File: verify/github_test.go
package verify

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"net/http"
	"testing"
)

func computeGitHubSignature(secret, body string) string {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(body))
	return "sha256=" + hex.EncodeToString(mac.Sum(nil))
}

func TestGitHubVerifyValidSignature(t *testing.T) {
	secret := "test-secret"
	body := `{"action":"push"}`
	sig := computeGitHubSignature(secret, body)

	v := NewGitHubVerifier(secret)
	headers := http.Header{}
	headers.Set("X-Hub-Signature-256", sig)

	ok, err := v.Verify([]byte(body), headers)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !ok {
		t.Fatal("expected valid signature to pass verification")
	}
}

func TestGitHubVerifyInvalidSignature(t *testing.T) {
	secret := "test-secret"
	body := `{"action":"push"}`
	wrongSig := computeGitHubSignature("wrong-secret", body)

	v := NewGitHubVerifier(secret)
	headers := http.Header{}
	headers.Set("X-Hub-Signature-256", wrongSig)

	ok, err := v.Verify([]byte(body), headers)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if ok {
		t.Fatal("expected invalid signature to fail verification")
	}
}

func TestGitHubVerifyMissingHeader(t *testing.T) {
	v := NewGitHubVerifier("test-secret")
	headers := http.Header{}

	_, err := v.Verify([]byte(`{}`), headers)
	if err == nil {
		t.Fatal("expected error for missing header")
	}
}

File: src/verify/github.test.ts
import { describe, it, expect } from "vitest";
import { createHmac } from "node:crypto";
import { GitHubVerifier } from "./github.js";

/** Computes a valid GitHub webhook signature for testing. */
function computeGitHubSignature(secret: string, body: string): string {
  const mac = createHmac("sha256", secret).update(body).digest("hex");
  return `sha256=${mac}`;
}

describe("GitHubVerifier", () => {
  it("accepts a valid signature", () => {
    const secret = "test-secret";
    const body = '{"action":"push"}';
    const sig = computeGitHubSignature(secret, body);

    const verifier = new GitHubVerifier(secret);
    const headers = { "x-hub-signature-256": sig };

    const result = verifier.verify(Buffer.from(body), headers);
    expect(result.error).toBeUndefined();
    expect(result.valid).toBe(true);
  });

  it("rejects an invalid signature", () => {
    const secret = "test-secret";
    const body = '{"action":"push"}';
    const wrongSig = computeGitHubSignature("wrong-secret", body);

    const verifier = new GitHubVerifier(secret);
    const headers = { "x-hub-signature-256": wrongSig };

    const result = verifier.verify(Buffer.from(body), headers);
    expect(result.error).toBeUndefined();
    expect(result.valid).toBe(false);
  });

  it("returns error for missing header", () => {
    const verifier = new GitHubVerifier("test-secret");
    const result = verifier.verify(Buffer.from("{}"), {});

    expect(result.error).toBeDefined();
    expect(result.valid).toBe(false);
  });
});
